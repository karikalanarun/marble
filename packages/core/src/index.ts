export { EffectFactory } from './effects/effects.factory';
export { defaultError$ } from './error/error.effect';
export { coreErrorFactory, CoreErrorOptions } from './error/error.factory';
export { HttpError, CoreError, EventError } from './error/error.model';
export { combineRoutes } from './router/router.combiner';
export { r } from './router/router.ixbuilder';
export { combineEffects, combineMiddlewares } from './effects/effects.combiner';
export { createEffectMetadata } from './effects/effectsMetadata.factory';
export * from './effects/effects.interface';
export * from './effects/http-effects.interface';
export * from './router/router.interface';
export * from './operators';
export * from './http.interface';
export * from './event/event.factory';
export * from './event/event.interface';
export * from './context/context.factory';
export * from './context/context.token.factory';
export * from './server/server.event';
export * from './server/server.interface';
export * from './server/server.tokens';
export { createServer } from './server/server.factory';
export { HttpListenerConfig, httpListener } from './listener/http.listener';
