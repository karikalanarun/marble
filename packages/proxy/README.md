<p align="center">
  <a href="https://marblejs.com">
    <img src="https://github.com/marblejs/marble/blob/master/assets/img/logo.png?raw=true" width="200" alt="Marble.js logo"/>
  </a>
</p>

# @marblejs/proxy

A simple proxy server for [Marble.js](https://github.com/marblejs/marble). It allows to send and receive requests from any server defined as `(req: IncomingMessage, res: OutgoingMessage) => void`. It is a base for `@marblejs/testing` and `@marblejs/serverless` packages.

## Installation

```
$ npm i @marblejs/proxy
```

## Documentation

For the latest updates, documentation, change log, and release information visit [docs.marblejs.com](https://docs.marblejs.com) and follow [@marble_js](https://twitter.com/marble_js) on Twitter.

## Usage

```typescript
import { ServerProxy, ServerApp, ServerProxyRequest, ServerProxyResponse } from '@marblejs/proxy';

export interface MyRequest {
  headers: Record<string, string>;
  method: string;
  path: string;
  body?: string;
}

export interface MyResponse {
  headers: Record<string, string>;
  body: string;
  code: number;
}

export class MyProxy extends ServerProxy<MyRequest, MyResponse> {

  constructor(app: ServerApp) {
    super(app);
  }

  normalizeError(error: Error): MyResponse {
    return {
      code: 502,
      body: error.message,
      headers: {},
    };
  }

  normalizeRequest(proxyRequest: MyRequest): ServerProxyRequest {
    const {
      body,
      headers,
      method,
      path,
    } = proxyRequest;
    return {
      headers,
      // Convert body to buffer
      body: body !== undefined ? Buffer.from(body) : undefined,
      method,
      path
    };
  }

  normalizeResponse(serverProxyResponse: ServerProxyResponse): MyResponse {
    const {
      headers,
      body,
      statusCode,
    } = serverProxyResponse;

    return {
      // Convert multi-value headers from Record<string, string[]>
      headers: Object.keys(headers).reduce((result, key) => {
        result[key] = headers[key][0];
        return headers;
      }, {}),
      // Convert body from Buffer to string
      body: body.toString(),
      code: statusCode,
    };
  }
}

```

License: MIT
