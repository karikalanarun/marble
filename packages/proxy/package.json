{
  "name": "@marblejs/proxy",
  "version": "2.2.2",
  "description": "A generic proxy server for Marble.js",
  "keywords": [
    "marble.js",
    "proxy",
    "server"
  ],
  "author": "Krzysztof Miemiec <krzysztof.miemiec@gmail.com>",
  "homepage": "https://github.com/marblejs/marble/tree/master/packages/proxy#readme",
  "license": "MIT",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist/"
  ],
  "publishConfig": {
    "access": "public"
  },
  "peerDependencies": {
    "@marblejs/middleware-body": "^2.2.2"
  },
  "devDependencies": {
    "@types/node": "^12.0.4",
    "@marblejs/core": "^2.2.2",
    "@marblejs/middleware-body": "^2.2.2"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/marblejs/marble.git"
  },
  "scripts": {
    "start": "yarn watch",
    "watch": "tsc -w",
    "build": "tsc",
    "clean": "rimraf dist",
    "test": "jest --config ../../jest.config.js"
  },
  "bugs": {
    "url": "https://github.com/marblejs/marble/issues"
  }
}
