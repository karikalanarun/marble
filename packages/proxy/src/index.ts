export * from './binaryMimeTypes';
export * from './serverProxy.helpers';
export * from './serverProxy';
