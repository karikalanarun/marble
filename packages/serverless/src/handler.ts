import { AwsLambdaProxyOptions } from './+awsLambda/awsLambda.types';
import { BoundDependency, createContext, httpListener, registerAll } from '@marblejs/core';

export enum ProxyType {
  AWS = 'aws',
}

type ProxyOptions<Type extends ProxyType> =
  Type extends ProxyType.AWS ? Partial<AwsLambdaProxyOptions> :
    never;

export interface LambdaOptions<Type extends ProxyType> {
  httpListener: ReturnType<typeof httpListener>;
  type: ProxyType;
  dependencies?: BoundDependency<any>[];
  proxyOptions?: ProxyOptions<Type>;
}

export const createLambda = <Type extends ProxyType>({
  httpListener,
  type,
  dependencies = [],
  proxyOptions
}: LambdaOptions<Type>) => {
  const context = registerAll([...dependencies])(createContext());
  const app = httpListener.run(context);
  switch (type) {
    case ProxyType.AWS:
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      const { createAwsLambdaHandler } = require('./+awsLambda');
      return createAwsLambdaHandler(app, proxyOptions);
    default:
      throw new Error(`Invalid type specified. Expected: ${Object.values(ProxyType)}`);
  }
};
