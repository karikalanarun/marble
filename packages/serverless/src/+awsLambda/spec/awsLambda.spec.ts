import { httpListener } from '@marblejs/core';
import { APIGatewayProxyEvent, APIGatewayProxyResult, Context } from 'aws-lambda';
import { createHandler } from './awsLambda.spec-setup';
import { createLambda } from '../../handler';
import { getHeaderByKey } from '@marblejs/proxy';

describe('@marblejs/serverless - AWS Proxy integration', () => {
  let handler;
  beforeAll(() => {
    handler = createHandler();
  });

  afterAll(() => {
    handler.close();
  });

  test('POST / returns 200', async () => {
    const event: APIGatewayProxyEvent = await import('./awsLambda.postEvent.mock.json') as any;
    const context: Context = await import('./awsLambda.postContext.mock.json') as any;
    const response: APIGatewayProxyResult = await handler(event, context);
    expect(JSON.parse(response.body)).toMatchSnapshot();
    expect(response.statusCode).toBe(200);
  });

  test('can be called multiple times asynchronously', async () => {
    const event: APIGatewayProxyEvent = await import('./awsLambda.postEvent.mock.json') as any;
    const context: Context = await import('./awsLambda.postContext.mock.json') as any;
    const eventRequest = async () => {
      const response = await handler(event, context);
      expect(response.statusCode).toBe(200);
    };
    await Promise.all(new Array(10).fill(null).map(eventRequest));
  });

  test('creates proxy wrapper only for AWS', () => {
    expect(() => createLambda({
      type: 'unknown' as any,
      httpListener: httpListener({ effects: [] }),
    })).toThrowError('Invalid type specified.');
  });

  test('removes "transfer-encoding: chunked" header', async () => {
    const response: APIGatewayProxyResult = await handler({
      httpMethod: 'GET',
      path: '/chunked-transfer',
    }, {});

    expect(response.statusCode).toBe(200);
    expect(response.multiValueHeaders).toBeDefined();
    expect(getHeaderByKey(response.multiValueHeaders, 'Content-Type')).toBe('application/json');
    expect(getHeaderByKey(response.multiValueHeaders, 'Content-Length')).toBe('15');
    expect(JSON.parse(response.body)).toEqual({ key: 'value' });
  });
});
