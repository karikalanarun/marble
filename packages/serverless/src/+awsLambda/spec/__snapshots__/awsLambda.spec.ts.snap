// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@marblejs/serverless - AWS Proxy integration POST / returns 200 1`] = `
Object {
  "body": "{\\"test\\":\\"value\\"}",
  "context": Object {
    "awsRequestId": "c482b12a-38bb-48bf-83a7-d7d8576dd442",
    "callbackWaitsForEmptyEventLoop": true,
    "functionName": "sample-lambda-1234567890AB",
    "functionVersion": "1",
    "invokedFunctionArn": "arn:aws:lambda:eu-central-1:123456789012:function:sample-lambda-1234567890AB:latest",
    "logGroupName": "/aws/lambda/sample-lambda-1234567890AB",
    "logStreamName": "2019/06/01/[latest]d89065cfaf354bd2ba025b5860ecb67e",
    "memoryLimitInMB": "128",
  },
  "event": Object {
    "httpMethod": "POST",
    "isBase64Encoded": false,
    "multiValueHeaders": Object {
      "Accept": Array [
        "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
      ],
      "Accept-Encoding": Array [
        "br, gzip, deflate",
      ],
      "Accept-Language": Array [
        "pl-pl",
      ],
      "CloudFront-Forwarded-Proto": Array [
        "https",
      ],
      "CloudFront-Is-Desktop-Viewer": Array [
        "true",
      ],
      "CloudFront-Is-Mobile-Viewer": Array [
        "false",
      ],
      "CloudFront-Is-SmartTV-Viewer": Array [
        "false",
      ],
      "CloudFront-Is-Tablet-Viewer": Array [
        "false",
      ],
      "CloudFront-Viewer-Country": Array [
        "PL",
      ],
      "Content-Type": Array [
        "text/plain",
      ],
      "Host": Array [
        "1234567890.execute-api.eu-central-1.amazonaws.com",
      ],
      "User-Agent": Array [
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_5) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.1.1 Safari/605.1.15",
      ],
      "Via": Array [
        "2.0 3095e870e1a1a1b03178e40ab1872de5.cloudfront.net (CloudFront)",
      ],
      "X-Amz-Cf-Id": Array [
        "Y25j7-itoeNCvCMwAt5YVWgDduoALWzQ_eL8dYb4-u2M5HC4VTyfeA==",
      ],
      "X-Amzn-Trace-Id": Array [
        "Root=1-5cf2885f-fa32290824a86664bc211d02",
      ],
      "X-Forwarded-For": Array [
        "127.0.0.1, 70.132.34.142",
      ],
      "X-Forwarded-Port": Array [
        "443",
      ],
      "X-Forwarded-Proto": Array [
        "https",
      ],
    },
    "multiValueQueryStringParameters": null,
    "path": "/",
    "pathParameters": null,
    "queryStringParameters": null,
    "requestContext": Object {
      "accountId": "534774995974",
      "apiId": "1234567890",
      "domainName": "1234567890.execute-api.eu-central-1.amazonaws.com",
      "domainPrefix": "1234567890",
      "extendedRequestId": "amo-8HzPliAFgHg=",
      "httpMethod": "POST",
      "identity": Object {
        "accessKey": null,
        "accountId": null,
        "caller": null,
        "cognitoAuthenticationProvider": null,
        "cognitoAuthenticationType": null,
        "cognitoIdentityId": null,
        "cognitoIdentityPoolId": null,
        "principalOrgId": null,
        "sourceIp": "127.0.0.1",
        "user": null,
        "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_5) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.1.1 Safari/605.1.15",
        "userArn": null,
      },
      "path": "/Prod/",
      "protocol": "HTTP/1.1",
      "requestId": "a17a92fe-8477-11e9-b7d2-51d9bcca17bd",
      "requestTime": "01/Jun/2019:14:14:55 +0000",
      "requestTimeEpoch": 1559398495665,
      "resourceId": "lm3hfkkof6",
      "resourcePath": "/",
      "stage": "Prod",
    },
    "resource": "/",
    "stageVariables": null,
  },
}
`;
